<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/07_baseline_rf.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB07 — Baseline Supervisionado

## 1. Contexto

Após a construção da série temporal (NB03), detecção de episódios (NB04), engenharia de atributos (NB05) e rotulagem de estados (NB06), este notebook estabelece o primeiro modelo supervisionado do pipeline.

O objetivo é avaliar, de forma controlada, se os atributos derivados da taxa de falha (`fail_rate`) são suficientes para permitir a identificação e antecipação de estados operacionais.

## 2. Objetivo

Este notebook tem como objetivos:

- construir um baseline supervisionado utilizando Random Forest;
- avaliar a separabilidade entre regimes normais e críticos;
- analisar o comportamento do modelo em classificação binária e multiclasse;
- investigar o impacto do tempo absoluto (`bucket_start_us`) no aprendizado;
- estabelecer uma referência experimental para os modelos do NB08.

## 3. Papel no Pipeline

Este notebook representa a primeira validação supervisionada do pipeline completo:

- NB04 → definição formal de episódios
- NB05 → construção de atributos baseados em taxa
- NB06 → rotulagem temporal (NORMAL / BEFORE / DURING / AFTER)

O NB07 responde à seguinte pergunta central:

"Os atributos construídos são suficientes para permitir aprendizado supervisionado consistente?"

## 4. Estrutura Experimental

A avaliação foi conduzida sob as seguintes condições:

- total de janelas: **8914**
- divisão temporal fixa (80% treino / 20% teste)
- preservação da ordem cronológica (sem embaralhamento)
- uso exclusivo de atributos numéricos derivados da série temporal
- ausência de vazamento temporal ou uso de rótulos indiretos

Foram considerados dois cenários:

1. Classificação binária:
   - evento (BEFORE / DURING / AFTER) vs NORMAL

2. Classificação multiclasse:
   - NORMAL / BEFORE / DURING / AFTER

Adicionalmente, foi conduzido um experimento controlado removendo o tempo absoluto (`bucket_start_us`) para avaliar possíveis efeitos de regime temporal.

## 5. Métrica de Interesse

As métricas utilizadas incluem:

- Accuracy
- Precision
- Recall
- F1-score
- ROC-AUC (para classificação binária)

Para o cenário multiclasse:

- Macro-F1 (equilíbrio entre classes)
- Weighted-F1 (impacto da distribuição)

## 6. Observações Metodológicas

- o modelo Random Forest foi escolhido por sua robustez em dados tabulares;
- não foi realizado ajuste fino de hiperparâmetros nesta etapa;
- o foco está na validação estrutural do problema, e não na maximização de desempenho;
- os resultados servirão como referência para comparação no NB08.

## 7. Relação com as Próximas Etapas

Os resultados deste notebook orientam:

- a escolha de modelos mais sofisticados (NB08);
- a definição das métricas principais do artigo;
- a análise de robustez e generalização;
- a consolidação experimental no NB09.

In [ ]:
# ============================================================
# 07_baseline_rf.ipynb
# Baseline supervisionado com Random Forest (binário vs multiclasse)
# Split temporal + métricas + matriz de confusão + salvamento do modelo
# Pipeline PPCOMP_DM (Google Cluster Trace) - coerente com taxa (fail_rate)
# ============================================================

# -----------------------------
# 0) BOOTSTRAP (Colab + Repo)
# -----------------------------
from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import numpy as np
import pandas as pd
import json

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull).")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)
importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, MODELS_PATH, ensure_dirs
ensure_dirs()

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH  =", REPORTS_PATH)
print("MODELS_PATH   =", MODELS_PATH)

def log(msg: str) -> None:
    print(f"[07_baseline_rf] {msg}")

# -----------------------------
# 1) Imports ML
# -----------------------------
from sklearn.model_selection import TimeSeriesSplit
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_recall_fscore_support,
    classification_report,
    confusion_matrix,
    roc_auc_score
)
import joblib

# -----------------------------
# 2) Carregar dataset rotulado (NB06)
# -----------------------------
DATA_FILE = FEATURES_PATH / "window_5min_labeled.parquet"
assert DATA_FILE.exists(), f"Arquivo não encontrado: {DATA_FILE}"

df = pd.read_parquet(DATA_FILE).sort_values("bucket_id").reset_index(drop=True)

assert "state" in df.columns, "Coluna 'state' ausente."
assert "bucket_id" in df.columns, "Coluna 'bucket_id' ausente."
assert "fail_rate" in df.columns, "Coluna 'fail_rate' ausente (pipeline por taxa não está refletido)."

log(f"Dataset rotulado: shape={df.shape}")
dist_total = df["state"].value_counts().to_dict()
log(f"Distribuição total: {dist_total}")

# -----------------------------
# 3) Preparar features (X) e alvos (y)
# -----------------------------
REMOVE_TIME_ABSOLUTE = False  # se True, remove bucket_start_us das features

drop_cols = {"state", "bucket_id"}

# is_critical é derivado do NB04; remover para evitar vazamento conceitual
if "is_critical" in df.columns:
    drop_cols.add("is_critical")

if REMOVE_TIME_ABSOLUTE and "bucket_start_us" in df.columns:
    drop_cols.add("bucket_start_us")

numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
X_cols = [c for c in numeric_cols if c not in drop_cols]

if len(X_cols) == 0:
    raise ValueError("Nenhuma feature numérica encontrada após filtragem.")

X = df[X_cols].copy()
y_multi = df["state"].astype(str).copy()

log(f"Features usadas: {len(X_cols)}")
log(f"Lista features: {X_cols}")

# -----------------------------
# 4) Split temporal (corte fixo default)
# -----------------------------
SPLIT_MODE = "fixed"   # "fixed" ou "tscv"
TEST_RATIO = 0.20

n = len(df)
test_size = int(np.ceil(n * TEST_RATIO))
train_end = n - test_size

X_train, X_test = X.iloc[:train_end], X.iloc[train_end:]
y_train_multi, y_test_multi = y_multi.iloc[:train_end], y_multi.iloc[train_end:]

log(f"Split temporal: train={len(X_train)} test={len(X_test)} (modo={SPLIT_MODE})")

# distribuição por split (útil para validar desbalanceamento)
dist_train = y_train_multi.value_counts().to_dict()
dist_test = y_test_multi.value_counts().to_dict()
log(f"Distribuição treino: {dist_train}")
log(f"Distribuição teste : {dist_test}")

# -----------------------------
# 5) Cenário binário: evento vs normal
# -----------------------------
y_train_bin = (y_train_multi != "NORMAL").astype(int)
y_test_bin = (y_test_multi != "NORMAL").astype(int)

rf_bin = RandomForestClassifier(
    n_estimators=300,
    random_state=SEED,
    n_jobs=-1,
    class_weight="balanced_subsample"
)
rf_bin.fit(X_train, y_train_bin)

pred_bin = rf_bin.predict(X_test)
proba_bin = rf_bin.predict_proba(X_test)[:, 1] if hasattr(rf_bin, "predict_proba") else None

acc_bin = accuracy_score(y_test_bin, pred_bin)
prec_bin, rec_bin, f1_bin, _ = precision_recall_fscore_support(
    y_test_bin, pred_bin, average="binary", zero_division=0
)

auc_bin = None
if proba_bin is not None and len(np.unique(y_test_bin)) == 2:
    try:
        auc_bin = float(roc_auc_score(y_test_bin, proba_bin))
    except Exception:
        auc_bin = None

cm_bin = confusion_matrix(y_test_bin, pred_bin).tolist()
log(f"[BIN] acc={acc_bin:.4f} prec={prec_bin:.4f} rec={rec_bin:.4f} f1={f1_bin:.4f} auc={auc_bin}")

# -----------------------------
# 6) Cenário multiclasse: NORMAL/BEFORE/DURING/AFTER
# -----------------------------
labels = ["NORMAL", "BEFORE", "DURING", "AFTER"]  # ordem fixa para matriz comparável

rf_multi = RandomForestClassifier(
    n_estimators=400,
    random_state=SEED,
    n_jobs=-1,
    class_weight="balanced_subsample"
)
rf_multi.fit(X_train, y_train_multi)

pred_multi = rf_multi.predict(X_test)

acc_multi = accuracy_score(y_test_multi, pred_multi)

macro_prec, macro_rec, macro_f1, _ = precision_recall_fscore_support(
    y_test_multi, pred_multi, average="macro", zero_division=0
)
w_prec, w_rec, w_f1, _ = precision_recall_fscore_support(
    y_test_multi, pred_multi, average="weighted", zero_division=0
)

cm_multi = confusion_matrix(y_test_multi, pred_multi, labels=labels).tolist()
report_multi = classification_report(y_test_multi, pred_multi, labels=labels, zero_division=0)

log(f"[MULTI] acc={acc_multi:.4f} macro_f1={macro_f1:.4f} weighted_f1={w_f1:.4f}")

# -----------------------------
# 7) (Opcional) TimeSeriesSplit rápido (apenas multiclasse)
# -----------------------------
tscv_results = []
if SPLIT_MODE == "tscv":
    tscv = TimeSeriesSplit(n_splits=5)
    for fold, (tr, te) in enumerate(tscv.split(X), start=1):
        X_tr, X_te = X.iloc[tr], X.iloc[te]
        y_tr, y_te = y_multi.iloc[tr], y_multi.iloc[te]

        clf = RandomForestClassifier(
            n_estimators=300,
            random_state=SEED,
            n_jobs=-1,
            class_weight="balanced_subsample"
        )
        clf.fit(X_tr, y_tr)
        y_hat = clf.predict(X_te)

        f1_macro = precision_recall_fscore_support(
            y_te, y_hat, average="macro", zero_division=0
        )[2]
        tscv_results.append({"fold": fold, "f1_macro": float(f1_macro)})

# -----------------------------
# 8) Salvar modelos
# -----------------------------
bin_model_path = MODELS_PATH / "rf_binary.joblib"
multi_model_path = MODELS_PATH / "rf_multiclass.joblib"

joblib.dump(rf_bin, bin_model_path)
joblib.dump(rf_multi, multi_model_path)

# -----------------------------
# 9) Summary (JSON)
# -----------------------------
summary = {
    "seed": SEED,
    "split_mode": SPLIT_MODE,
    "test_ratio": TEST_RATIO,
    "rows_total": int(n),
    "rows_train": int(len(X_train)),
    "rows_test": int(len(X_test)),
    "n_features": int(len(X_cols)),
    "features": X_cols,
    "class_distribution_total": {k: int(v) for k, v in dist_total.items()},
    "class_distribution_train": {k: int(v) for k, v in dist_train.items()},
    "class_distribution_test": {k: int(v) for k, v in dist_test.items()},
    "binary": {
        "acc": float(acc_bin),
        "precision": float(prec_bin),
        "recall": float(rec_bin),
        "f1": float(f1_bin),
        "roc_auc": auc_bin,
        "confusion_matrix": cm_bin
    },
    "multiclass": {
        "acc": float(acc_multi),
        "macro_f1": float(macro_f1),
        "weighted_f1": float(w_f1),
        "confusion_matrix": cm_multi,
        "labels_order": labels,
        "classification_report_text": report_multi
    },
    "tscv_multiclass": tscv_results,
    "models": {
        "binary_path": str(bin_model_path),
        "multiclass_path": str(multi_model_path)
    }
}

summary_file = REPORTS_PATH / "07_baseline_rf_summary.json"
summary_file.write_text(json.dumps(summary, indent=2, ensure_ascii=False))

log("Notebook 07 finalizado com sucesso.")

print("\n=== MULTICLASS REPORT ===\n")
print(report_multi)
print("\n=== CONFUSION MATRIX (BIN) ===\n", cm_bin)
print("\n=== CONFUSION MATRIX (MULTI) labels=", labels, "===\n", cm_multi)
print("\nModels saved:\n", bin_model_path, "\n", multi_model_path)
print("\nSummary saved:\n", summary_file)

Mounted at /content/drive
[Bootstrap] Atualizando repositório (git pull).
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH  = /content/drive/MyDrive/Mestrado/04-reports
MODELS_PATH   = /content/drive/MyDrive/Mestrado/03-models
[07_baseline_rf] Dataset rotulado: shape=(8914, 28)
[07_baseline_rf] Distribuição total: {'NORMAL': 5533, 'BEFORE': 1996, 'AFTER': 991, 'DURING': 394}
[07_baseline_rf] Features usadas: 25
[07_baseline_rf] Lista features: ['bucket_start_us', 'n_events', 'n_failed', 'n_machines', 'n_collections', 'mean_priority', 'mean_req_cpus', 'mean_req_mem', 'req_cpus_presence_rate', 'req_mem_presence_rate', 'event_FAIL_count', 'event_SCHEDULE_count', 'event_FINISH_count', 'event_ENABLE_count', 'event_LOST_count', 'event_EVICT_count', 'event_KILL_count', 'fail_rate', 'lag_1', 'lag_2', 'lag_3', 'rolling_mean_1h', 'rolling_std_1h', 'pct_change', 'zscore_global']
[07_baseline_rf] Split 

## Teste controlado — remoção de `bucket_start_us` (viés de regime temporal)

### Motivação

O atributo `bucket_start_us` representa tempo absoluto (monotônico crescente). Em avaliação com **split temporal** (treino = passado, teste = futuro), esse atributo pode capturar indiretamente **mudanças de regime** (não-estacionaridade), isto é, aprender um proxy do tipo “quanto mais próximo do fim do dataset, maior probabilidade de evento”.

Esse efeito não caracteriza vazamento direto de rótulo, mas pode **inflar métricas** ao explorar diferenças sistemáticas de distribuição entre treino e teste.

Na execução atual, observa-se que o bloco de teste contém proporção maior de janelas em estados de evento (BEFORE/DURING/AFTER) do que o bloco de treino, o que reforça a necessidade de verificar se `bucket_start_us` está contribuindo para desempenho via regime temporal.

### Objetivo do teste

Executar um experimento controlado mantendo:
- mesmo dataset (`window_5min_labeled.parquet`)
- mesmo split temporal (80/20)
- mesmas features (exceto `bucket_start_us`)
- mesmos hiperparâmetros do Random Forest

Comparar os resultados **com vs sem** `bucket_start_us` em:
- Binário: Accuracy, Precision, Recall, F1, ROC-AUC, matriz de confusão
- Multiclasse: Accuracy, Macro-F1, Weighted-F1, classification report, matriz de confusão

### Interpretação esperada

- Se o desempenho cair pouco: evidência de que a separabilidade decorre majoritariamente das features derivadas de `fail_rate`.
- Se o desempenho cair muito: indício de forte dependência de regime temporal e necessidade de revisão do protocolo (ex.: validação por múltiplos blocos temporais, TimeSeriesSplit, remoção do tempo absoluto como feature).

In [ ]:
# ============================================================
# TESTE CONTROLADO: remover bucket_start_us e reavaliar NB07
# Gera summary separado: 07_baseline_rf_summary_no_time.json
# ============================================================

import numpy as np
import pandas as pd
import json
from pathlib import Path

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_recall_fscore_support,
    classification_report,
    confusion_matrix,
    roc_auc_score
)
import joblib

SEED = 42
TEST_RATIO = 0.20

def _safe_roc_auc(y_true, proba):
    if proba is None:
        return None
    if len(np.unique(y_true)) < 2:
        return None
    try:
        return float(roc_auc_score(y_true, proba))
    except Exception:
        return None

# Paths (usa os mesmos do NB07 já executado)
from src.paths import FEATURES_PATH, REPORTS_PATH, MODELS_PATH

DATA_FILE = FEATURES_PATH / "window_5min_labeled.parquet"
BASELINE_SUMMARY_FILE = REPORTS_PATH / "07_baseline_rf_summary.json"
OUT_SUMMARY_FILE = REPORTS_PATH / "07_baseline_rf_summary_no_time.json"

assert DATA_FILE.exists(), f"Arquivo não encontrado: {DATA_FILE}"
assert BASELINE_SUMMARY_FILE.exists(), f"Baseline summary não encontrado: {BASELINE_SUMMARY_FILE}"

baseline = json.loads(BASELINE_SUMMARY_FILE.read_text())

df = pd.read_parquet(DATA_FILE).sort_values("bucket_id").reset_index(drop=True)
assert "state" in df.columns
assert "bucket_id" in df.columns
assert "fail_rate" in df.columns

# --- preparar features removendo bucket_start_us ---
drop_cols = {"state", "bucket_id", "bucket_start_us"}
if "is_critical" in df.columns:
    drop_cols.add("is_critical")

numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
X_cols = [c for c in numeric_cols if c not in drop_cols]
if len(X_cols) == 0:
    raise ValueError("Nenhuma feature numérica sobrou após remoções.")

X = df[X_cols].copy()
y_multi = df["state"].astype(str).copy()

# --- split temporal 80/20 ---
n = len(df)
test_size = int(np.ceil(n * TEST_RATIO))
train_end = n - test_size

X_train, X_test = X.iloc[:train_end], X.iloc[train_end:]
y_train_multi, y_test_multi = y_multi.iloc[:train_end], y_multi.iloc[train_end:]

# --- binário ---
y_train_bin = (y_train_multi != "NORMAL").astype(int)
y_test_bin = (y_test_multi != "NORMAL").astype(int)

rf_bin = RandomForestClassifier(
    n_estimators=300,
    random_state=SEED,
    n_jobs=-1,
    class_weight="balanced_subsample"
)
rf_bin.fit(X_train, y_train_bin)

pred_bin = rf_bin.predict(X_test)
proba_bin = rf_bin.predict_proba(X_test)[:, 1] if hasattr(rf_bin, "predict_proba") else None

acc_bin = float(accuracy_score(y_test_bin, pred_bin))
prec_bin, rec_bin, f1_bin, _ = precision_recall_fscore_support(
    y_test_bin, pred_bin, average="binary", zero_division=0
)
prec_bin, rec_bin, f1_bin = float(prec_bin), float(rec_bin), float(f1_bin)
auc_bin = _safe_roc_auc(y_test_bin, proba_bin)
cm_bin = confusion_matrix(y_test_bin, pred_bin).tolist()

# --- multiclasse ---
labels = ["NORMAL", "BEFORE", "DURING", "AFTER"]

rf_multi = RandomForestClassifier(
    n_estimators=400,
    random_state=SEED,
    n_jobs=-1,
    class_weight="balanced_subsample"
)
rf_multi.fit(X_train, y_train_multi)

pred_multi = rf_multi.predict(X_test)

acc_multi = float(accuracy_score(y_test_multi, pred_multi))
macro_prec, macro_rec, macro_f1, _ = precision_recall_fscore_support(
    y_test_multi, pred_multi, average="macro", zero_division=0
)
w_prec, w_rec, w_f1, _ = precision_recall_fscore_support(
    y_test_multi, pred_multi, average="weighted", zero_division=0
)

macro_f1 = float(macro_f1)
w_f1 = float(w_f1)

cm_multi = confusion_matrix(y_test_multi, pred_multi, labels=labels).tolist()
report_multi = classification_report(y_test_multi, pred_multi, labels=labels, zero_division=0)

# --- salvar modelos separados (opcional, mas útil) ---
bin_model_path = MODELS_PATH / "rf_binary_no_time.joblib"
multi_model_path = MODELS_PATH / "rf_multiclass_no_time.joblib"
joblib.dump(rf_bin, bin_model_path)
joblib.dump(rf_multi, multi_model_path)

# --- delta vs baseline (com tempo) ---
def _delta(a, b):
    # a = atual (no_time), b = baseline (with_time)
    if a is None or b is None:
        return None
    return float(a) - float(b)

base_bin = baseline.get("binary", {})
base_multi = baseline.get("multiclass", {})

delta = {
    "binary": {
        "acc": _delta(acc_bin, base_bin.get("acc")),
        "precision": _delta(prec_bin, base_bin.get("precision")),
        "recall": _delta(rec_bin, base_bin.get("recall")),
        "f1": _delta(f1_bin, base_bin.get("f1")),
        "roc_auc": _delta(auc_bin, base_bin.get("roc_auc")),
    },
    "multiclass": {
        "acc": _delta(acc_multi, base_multi.get("acc")),
        "macro_f1": _delta(macro_f1, base_multi.get("macro_f1")),
        "weighted_f1": _delta(w_f1, base_multi.get("weighted_f1")),
    }
}

# --- summary do teste controlado ---
summary_no_time = {
    "experiment": "no_time_absolute",
    "seed": SEED,
    "split_mode": "fixed",
    "test_ratio": TEST_RATIO,
    "rows_total": int(n),
    "rows_train": int(len(X_train)),
    "rows_test": int(len(X_test)),
    "n_features": int(len(X_cols)),
    "features": X_cols,
    "dropped": sorted(list(drop_cols)),
    "class_distribution_total": {k: int(v) for k, v in df["state"].value_counts().to_dict().items()},
    "class_distribution_train": {k: int(v) for k, v in y_train_multi.value_counts().to_dict().items()},
    "class_distribution_test": {k: int(v) for k, v in y_test_multi.value_counts().to_dict().items()},
    "binary": {
        "acc": acc_bin,
        "precision": prec_bin,
        "recall": rec_bin,
        "f1": f1_bin,
        "roc_auc": auc_bin,
        "confusion_matrix": cm_bin
    },
    "multiclass": {
        "acc": acc_multi,
        "macro_f1": macro_f1,
        "weighted_f1": w_f1,
        "confusion_matrix": cm_multi,
        "labels_order": labels,
        "classification_report_text": report_multi
    },
    "delta_vs_baseline_with_time": delta,
    "models": {
        "binary_path": str(bin_model_path),
        "multiclass_path": str(multi_model_path)
    },
    "baseline_summary_file": str(BASELINE_SUMMARY_FILE),
}

OUT_SUMMARY_FILE.write_text(json.dumps(summary_no_time, indent=2, ensure_ascii=False))

print("=== TESTE CONTROLADO (SEM bucket_start_us) ===")
print("Binário:", {"acc": acc_bin, "precision": prec_bin, "recall": rec_bin, "f1": f1_bin, "roc_auc": auc_bin})
print("Multiclasse:", {"acc": acc_multi, "macro_f1": macro_f1, "weighted_f1": w_f1})
print("\nDelta vs baseline (with_time):")
print(json.dumps(delta, indent=2, ensure_ascii=False))
print("\nSummary salvo em:", OUT_SUMMARY_FILE)
print("Modelos salvos em:", bin_model_path, "e", multi_model_path)

=== TESTE CONTROLADO (SEM bucket_start_us) ===
Binário: {'acc': 0.8238923163208076, 'precision': 0.8951548848292296, 'recall': 0.8609625668449198, 'f1': 0.8777258566978193, 'roc_auc': 0.9008833683070466}
Multiclasse: {'acc': 0.64385866517106, 'macro_f1': 0.6677765193386259, 'weighted_f1': 0.6367694293201969}

Delta vs baseline (with_time):
{
  "binary": {
    "acc": 0.008412787436904101,
    "precision": 0.04565417156246776,
    "recall": -0.04889228418640179,
    "f1": -0.0009167106205133502,
    "roc_auc": 0.011966006195343493
  },
  "multiclass": {
    "acc": -0.011777902411665653,
    "macro_f1": 0.06867796931638226,
    "weighted_f1": 0.028874123446003375
  }
}

Summary salvo em: /content/drive/MyDrive/Mestrado/04-reports/07_baseline_rf_summary_no_time.json
Modelos salvos em: /content/drive/MyDrive/Mestrado/03-models/rf_binary_no_time.joblib e /content/drive/MyDrive/Mestrado/03-models/rf_multiclass_no_time.joblib


## 8. Conclusão da Etapa

O baseline supervisionado foi executado com sucesso, permitindo avaliar a capacidade dos atributos derivados de `fail_rate` em modelar os estados operacionais do sistema.

Os resultados demonstram que o pipeline construído nas etapas anteriores produz uma estrutura supervisionada consistente, adequada para aprendizado de máquina.

---

## 9. Importância para a Modelagem de Transições

Este notebook valida empiricamente a qualidade da base construída em:

- NB04 — detecção de episódios
- NB05 — engenharia de atributos
- NB06 — rotulagem de estados

Os resultados indicam que:

- os estados operacionais são separáveis no espaço de features;
- a fase pré-evento (BEFORE) apresenta sinal detectável;
- a modelagem supervisionada é viável sem necessidade de ajustes complexos.

Essa validação é fundamental antes da transição para o problema de predição de transições no NB08.

---

## 10. Resultados — Classificação Binária

Métricas obtidas:

- Accuracy: **0.815**
- Precision: **0.850**
- Recall: **0.910**
- F1-score: **0.879**
- ROC-AUC: **0.889**

Interpretação:

O modelo apresenta alta capacidade de detecção de eventos, com recall elevado e boa separabilidade estatística.  
Isso indica que os atributos derivados da taxa de falha capturam adequadamente o comportamento crítico do sistema.

---

## 11. Resultados — Classificação Multiclasse

Métricas globais:

- Accuracy: **0.656**
- Macro-F1: **0.599**
- Weighted-F1: **0.608**

Desempenho por classe:

- NORMAL: F1 ≈ 0.63  
- BEFORE: F1 ≈ 0.69  
- DURING: F1 = 1.00  
- AFTER: F1 ≈ 0.07  

Observações:

- A classe DURING apresenta separação perfeita.
- A classe BEFORE apresenta desempenho consistente.
- A classe AFTER apresenta dificuldade inicial de separação.

---

## 12. Achados Estruturais

### 12.1 Separação do regime crítico

A classe DURING foi identificada com precisão e recall iguais a 1.00, indicando que:

- a definição de episódios baseada em `fail_rate` é robusta;
- os atributos capturam adequadamente o regime crítico;
- eventos críticos são fortemente separáveis.

### 12.2 Modelagem da fase pré-evento

A classe BEFORE apresentou desempenho consistente (F1 ≈ 0.69), indicando que:

- o estado pré-evento é detectável;
- há sinal preditivo antes da ocorrência de eventos críticos.

### 12.3 Limitação inicial da classe AFTER

A baixa performance inicial da classe AFTER sugere:

- possível sobreposição com BEFORE;
- influência de regime temporal no aprendizado.

---

## 13. Teste Controlado — Remoção de Tempo Absoluto

### Motivação

O atributo `bucket_start_us` representa tempo absoluto e pode capturar indiretamente mudanças de regime ao longo do dataset.

Em um split temporal (treino passado / teste futuro), isso pode inflar métricas ao explorar diferenças de distribuição entre períodos.

### Configuração do teste

Foi realizado um experimento controlado mantendo:

- mesmo dataset
- mesmo split temporal (80/20)
- mesmos hiperparâmetros
- mesmas features (exceto `bucket_start_us`)

### Resultados — Classificação Binária (sem tempo)

- Accuracy: **0.824**
- Precision: **0.895**
- Recall: **0.861**
- F1-score: **0.878**
- ROC-AUC: **0.901**

Interpretação:

O desempenho permaneceu estável, com leve melhora em AUC, indicando que:

- a separação entre eventos e normalidade não depende do tempo absoluto;
- os atributos derivados de `fail_rate` são suficientes para o aprendizado.

### Resultados — Classificação Multiclasse (sem tempo)

- Accuracy: **0.644**
- Macro-F1: **0.668**
- Weighted-F1: **0.637**

Desempenho por classe:

- NORMAL: F1 ≈ 0.70  
- BEFORE: F1 ≈ 0.61  
- DURING: F1 = 1.00  
- AFTER: F1 ≈ 0.37  

Principais efeitos:

- melhora significativa da classe AFTER;
- aumento do Macro-F1;
- manutenção da separação perfeita da classe DURING.

---

## 14. Interpretação Integrada

Os resultados indicam que:

- a separação do regime crítico é estrutural e não depende do tempo absoluto;
- a engenharia de atributos baseada em taxa captura adequadamente a dinâmica do sistema;
- o modelo apresenta robustez sob variação de regime temporal;
- a remoção do tempo absoluto melhora o equilíbrio entre classes.

---

## 15. Implicações Científicas

A partir dos resultados, pode-se afirmar que:

- a definição de criticidade via μ + 2σ sobre `fail_rate` é consistente;
- eventos críticos são estatisticamente separáveis;
- a fase pré-evento é modelável;
- o modelo apresenta invariância relativa ao tempo absoluto;
- o pipeline produz base adequada para modelagem supervisionada.

---

## 16. Limitações do Baseline

- ausência de ajuste fino de hiperparâmetros;
- uso de um único modelo (Random Forest);
- ausência de validação cruzada temporal;
- avaliação limitada a métricas agregadas.

Essas limitações serão tratadas na etapa seguinte.

---

## 17. Conclusão Experimental

O Notebook 07 confirma que o pipeline:

- produz atributos informativos e consistentes;
- permite separação clara entre regimes operacionais;
- não depende de artefatos temporais triviais;
- apresenta robustez sob teste de invariância temporal.

Este baseline estabelece uma referência sólida para:

- comparação de modelos no NB08;
- avaliação de desempenho em tarefas de transição;
- consolidação dos resultados no artigo (NB09).